### PR TITLE
QHelp

### PR DESCRIPTION
**Pull request by arthurnn**
cc @marcogario

### CODE REVIEW

File: lib/code_scanning/qhelp_generator.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+require_relative '../code_scanning'
+
+class QHelpGenerator
```
Comment (marcogario) [inline lib/code_scanning/qhelp_generator.rb]: Minor: Technically qhelp is the name of the XML format used by codeQL. We call these "rules definitions" in TS. SarifHelp might also be a good name.

File: lib/code_scanning/rubocop/rule.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      @config ||= RuboCop::ConfigStore.new.for(Pathname.new(Dir.pwd))
+      @cop_config ||= @config.for_cop(@cop.department.to_s)
+                        .merge(@config.for_cop(@cop))
+    end
+
+    def to_h
+      properties = {
+        "id" => @cop_name,
+        "precision" => "very-high",
+        "problem.severity" => sarif_severity,
```
Comment (marcogario) [inline lib/code_scanning/rubocop/rule.rb]:
```suggestion

```

Can be removed. The proper way is to use the `defaultconfiguration` as you do below.

File: lib/code_scanning/rubocop/rule.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    def cop_config
+      @config ||= RuboCop::ConfigStore.new.for(Pathname.new(Dir.pwd))
+      @cop_config ||= @config.for_cop(@cop.department.to_s)
+                        .merge(@config.for_cop(@cop))
+    end
+
+    def to_h
+      properties = {
+        "id" => @cop_name,
```
Comment (marcogario) [inline lib/code_scanning/rubocop/rule.rb]:
```suggestion

```

Can be removed. The `id` is already defined at a higher level.

File: lib/code_scanning/rubocop/rule.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+          "text" => help,
+          "markdown" => help,
+        }
+        properties["queryURI"] = query_uri if badge.qualified?
+      end
+
+      if badge.qualified?
+        kind = badge.department.to_s
+        properties["tags"] = [kind.downcase]
+        properties["kind"] = kind
```
Comment (marcogario) [inline lib/code_scanning/rubocop/rule.rb]:
```suggestion

```

Can be removed. CodeQL specific.

File: lib/code_scanning/rubocop/rule.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      end
+
+       # TODO: add to properties
+#      {
+#        "name" => "Constant return type",
+#      }
```
Comment (marcogario) [inline lib/code_scanning/rubocop/rule.rb]: Left-over?

File: lib/code_scanning/rubocop/sarif_formatter.rb [hunk 6]
```diff
@@ -1,71 +1,47 @@
...
+      }
       @rules_map = {}
-      @results = []
       @rules = []
+      @results = []
       @sarif['runs'] = [
-        { 'tool' => {
+        {
+          'tool' => {
             'driver' => { 'name' => 'Rubocop', 'rules' => @rules }
```
Comment (marcogario) [inline lib/code_scanning/rubocop/sarif_formatter.rb:20]: Did you confirm that the `name` matches what you get in a SARIF file generated by the tool?

**Review by marcogario — COMMENTED** [review]
can you link a generated file too?

File: lib/code_scanning/rubocop/rule.rb (hunk 5 above)
Comment (arthurnn) [reply lib/code_scanning/rubocop/rule.rb]: Do we need the name on the property? I saw the codeql one has it

File: lib/code_scanning/rubocop/sarif_formatter.rb (hunk 6 above)
Comment (arthurnn) [reply lib/code_scanning/rubocop/sarif_formatter.rb:20]: This Sarif formatter code is used to generate the rules and the run sarif too

File: lib/code_scanning/rubocop/rule.rb (hunk 5 above)
Comment (marcogario) [reply lib/code_scanning/rubocop/rule.rb]: no, the name is only at the top level.

**Review by marcogario — APPROVED** [review]

File: lib/code_scanning/rubocop/sarif_formatter.rb (hunk 6 above)
Comment (marcogario) [reply lib/code_scanning/rubocop/sarif_formatter.rb:20]: Cool! I somehow thought this was outputing SARIF already. Great that you managed to wrap it!

**Pull request merged**